src/values.rs

```diff
@@ -3,8 +3,78 @@ use std::ops::Range;
 
 use bevy_reflect::{FromReflect, Reflect};
 use bevy_render::prelude::Color;
+use rand::seq::SliceRandom;
 use rand::{prelude::ThreadRng, Rng};
 
+/// A value that will be chosen from a set of possible values when read.
+///
+/// ## Examples
+///
+/// ``T`` values can be converted into ``Constant``
+/// [`Range<T>`]s or [`Vec<T>`]s can be converted into ``RandomChoice``
+///
+/// ## Examples
+/// ```
+/// # use bevy_particle_systems::values::{RandomValue};
+/// # use rand;
+///
+/// let mut rng = rand::thread_rng();
+///
+/// // Results in a constant value
+/// let c: RandomValue<usize> = (2).into();
+///
+/// // Results are picked randomly from a range
+/// let r: RandomValue<usize> = (1..3).into();
+///
+/// // Results are picked randomly from a set of values
+/// let v: RandomValue<usize> = vec![0, 2, 4, 8].into();
+/// ```
+
+#[derive(Debug, Clone, Reflect, FromReflect)]
+pub enum RandomValue<T: Reflect + Clone + FromReflect> {
+    /// A constant value
+    Constant(T),
+
+    /// A set of possible values to choose from randomly
+    RandomChoice(Vec<T>),
+}
+
+impl<T: Reflect + Clone + FromReflect> From<T> for RandomValue<T> {
+    fn from(t: T) -> Self {
+        RandomValue::Constant(t)
+    }
+}
+
+impl<T: Reflect + Clone + FromReflect> From<Range<T>> for RandomValue<T>
+where Range<T>: Iterator<Item = T> {
+    fn from(r: Range<T>) -> Self {
+        RandomValue::RandomChoice(r.collect())
+    }
+}
+
+impl<T: Reflect + Clone + FromReflect> From<Vec<T>> for RandomValue<T> {
+    fn from(v: Vec<T>) -> Self {
+        RandomValue::RandomChoice(v)
+    }
+}
+
+impl<T: Reflect + Clone + FromReflect> RandomValue<T> {
```

Comment:
I wonder if it would be better to use `Copy` vs `Clone` here. It would get rid of the two `.clone()` calls below, and in theory `T` should be something that's very cheap to copy anyway, or this will potentially substantially impact performance at high spawn rates (or with other future uses).

Reply:
Awesome!
I would lean towards using `Clone` here:
- if the user uses a `Copy` type, the `.clone()` call will just copy the bits so will be very cheap
- we still leave the freedom to the user of using `Copy` types, where they could benchmark the performance impact on their own.

Basically I think it's more freedom with no downsides

Reply:
Ah, fair point! Agreed.